# Gambler’s Ruin Problem

Consider a gambler who starts with an initial fortune of \$ 1 and then on each successive gamble either wins \$ 1 or loses \$ 1 independent of the past with probabilities p and q = 1−p respectively. Let Rn denote the total fortune after the nth gamble. The gambler’s objective is to reach a total fortune of \$ N, without first getting ruined (running out of money). If the gambler succeeds, then the gambler is said to win the game. In any case, the gambler stops playing after winning or getting ruined, whichever happens first. There is nothing special about starting with \$ 1, more generally the gambler starts with $i where 0 < i < N.

In [1]:
p = 0.4 # probability of winning
q = 0.6 # probability of losing
N = 5 # end state

In [2]:
# Form the transition matrix

import numpy as np

P = np.array([[1, 0, 0, 0, 0, 0],
                     [0.6, 0, 0.4, 0, 0, 0],
                     [0, 0.6, 0, 0.4, 0, 0],
                     [0, 0, 0.6, 0, 0.4, 0],
                     [0, 0, 0, 0.6, 0, 0.4],
                     [0, 0, 0, 0, 0, 1]])

\begin{bmatrix}
    1 & 0 & 0 & 0 & 0 & 0 \\
    0.6 & 0 & 0.4 & 0 & 0 & 0 \\
    0 & 0.6 & 0 & 0.4 & 0 & 0 \\
    0 & 0 & 0.6 & 0 & 0.4 & 0 \\
    0 & 0 & 0 & 0.6 & 0 & 0.4 \\
    0 & 0 & 0 & 0 & 0 & 1
\end{bmatrix}

## Solve
$$V(P-I) = 0$$

## Let
$$V = [a, b, c, d, e, f]$$

$$P-I = \begin{bmatrix}
    0 & 0 & 0 & 0 & 0 & 0 \\
    0.6 & -1 & 0.4 & 0 & 0 & 0 \\
    0 & 0.6 & -1 & 0.4 & 0 & 0 \\
    0 & 0 & 0.6 & -1 & 0.4 & 0 \\
    0 & 0 & 0 & 0.6 & -1 & 0.4 \\
    0 & 0 & 0 & 0 & 0 & 0
\end{bmatrix} $$

## Set of equations

$$\rightarrow a+b+c+d+e+f=1$$

$$\rightarrow b*0.6=0$$

$$\rightarrow b*-1 + c*0.6=0$$

$$\rightarrow b*0.4 + c*-1 + d*0.6=0$$

$$\rightarrow c*0.4 + d*-1 + e*0.6=0$$

$$\rightarrow d*0.4 + e*-1 =0$$

$$\rightarrow e*0.4=0$$

# Solutions
$$ a = x$$

$$ b = c = d = e = 0$$

$$ f = 1 - x$$

$$ V = [x, 0, 0, 0, 0, 1-x]$$

## Steady state vector V tells that probability of being in intermediate states is zero after large number of trials.

# Case 1

Let us start with \$ 2 initially.
Initial state vector:
$$ V = [0, 0, 1, 0, 0, 0]$$

$$ VP^{10000} = [0 0 1 0 0 0]\begin{bmatrix}
    1 & 0 & 0 & 0 & 0 & 0 \\
    0.6 & 0 & 0.4 & 0 & 0 & 0 \\
    0 & 0.6 & 0 & 0.4 & 0 & 0 \\
    0 & 0 & 0.6 & 0 & 0.4 & 0 \\
    0 & 0 & 0 & 0.6 & 0 & 0.4 \\
    0 & 0 & 0 & 0 & 0 & 1
\end{bmatrix}^{10000} $$

In [3]:
np.array([0, 0, 1, 0, 0, 0]).dot(np.linalg.matrix_power(P, 10000))

array([0.81042654, 0.        , 0.        , 0.        , 0.        ,
       0.18957346])

$$ VP^{10000} = [0.81042654, 0       , 0        , 0     , 0  ,0.18957346]$$

## So, if we start with $2, the probability of losing all money is 0.8105 and probability of winning is 0.1896 after large number of trials.

# Case 2

Let us start with \$ 4 initially.
Initial state vector:
$$ V = [0, 0, 0, 0, 1, 0]$$

$$ VP^{10000} = [0 0 0 0 1 0]\begin{bmatrix}
    1 & 0 & 0 & 0 & 0 & 0 \\
    0.6 & 0 & 0.4 & 0 & 0 & 0 \\
    0 & 0.6 & 0 & 0.4 & 0 & 0 \\
    0 & 0 & 0.6 & 0 & 0.4 & 0 \\
    0 & 0 & 0 & 0.6 & 0 & 0.4 \\
    0 & 0 & 0 & 0 & 0 & 1
\end{bmatrix}^{10000} $$

In [4]:
np.array([0, 0, 0, 0, 1, 0]).dot(np.linalg.matrix_power(P, 10000))

array([0.38388626, 0.        , 0.        , 0.        , 0.        ,
       0.61611374])

$$ VP^{10000} = [0.38388626, 0      , 0      , 0       , 0        ,0.61611374]$$

## So, if we start with $4, the probability of losing all money is 0.3839 and probability of winning is 0.6162 after large number of trials.